Added remove_duplicate parameter to `nn.Module`

Chillee opened the pull request:
This makes it so that shared parameters get their own entry in `named_parameters`.

More broadly, this makes it so that
```
params_and_buffers = {**mod.named_named_parameters(remove_duplicate=False), **mod.named_buffers(remove_duplicate=False)}
_stateless.functional_call(mod, params_and_buffers, args, kwargs)
```
is identical to calling the original module's forwards pass.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

jbschlosser reviewed (commented):
Thanks for the update! Functionality looks pretty good, but do you mind adding a test or two? Could tweak the existing tests or drop next to them: https://github.com/pytorch/pytorch/blob/9f0c8085932d37d01557363e03a25185cdf81cea/test/test_nn.py#L1069

albanD commented on torch/nn/modules/module.py:
```diff
@@ -1491,7 +1491,7 @@ def load(module, prefix=''):
+    def _named_members(self, get_members_fn, prefix='', recurse=True, remove_duplicate=True):
         r"""Helper method for yielding various names + members of modules."""
         memo = set()
         modules = self.named_modules(prefix=prefix) if recurse else [(prefix, self)]
```
You need to pass the remove_duplicate flag to the module traversal here no?

jbschlosser replied:
Good catch - would be good to make sure the case of duplicate modules is tested as well

Chillee commented on torch/nn/modules/module.py:
```diff
@@ -1537,6 +1538,9 @@ def named_parameters(self, prefix: str = '', recurse: bool = True) -> Iterator[T
                 are direct members of this module.
+            remove_duplicate (bool): if True, then removes parameters
+                that are duplicates of each other. For example, if two
+                parameters are tied, it'll only return one.
```
> I guess the first it encounters when traversing the module hierarchy
That's true, but I'm not totally sure that's very actionable from the user's end.

Chillee commented:
I'm not sure it's ... possible to have duplicate buffers? Is there an example of how that's possible?

Otherwise, added tests.

jbschlosser commented on test/test_nn.py:
```diff
@@ -1118,6 +1118,25 @@ def names(named_buffers):
+                self.linear.bias = self.bias
+                self.linear_cloned = self.linear
+                self.register_buffer('buffer', torch.randn(3))
+                self.linear.buffer_cloned = self.buffer
```
I think `self.register_buffer('buffer_cloned', self.buffer)` works for cloning a buffer

albanD commented on test/test_nn.py:
```diff
@@ -1118,6 +1118,24 @@ def names(named_buffers):
+                self.bias = nn.Parameter(torch.randn(3))
+                self.linear = nn.Linear(3, 3)
+                self.linear.bias = self.bias
+                self.linear_cloned = self.linear
```
I don't think "cloned" is a good word here. We are not actually cloning but aliasing

albanD reviewed (approved):
SGTM

jbschlosser reviewed (approved):
LGTM as well

Chillee committed:
Added remove_duplicate parameter to `nn.Module` (#39)

Summary:
Pull Request resolved: https://github.com/pytorch/torchrec/pull/39

Pull Request resolved: https://github.com/facebookresearch/torchrec/pull/6

This makes it so that shared parameters get their own entry in `named_parameters`.

More broadly, this makes it so that
```
params_and_buffers = {**mod.named_named_parameters(remove_duplicate=False), **mod.named_buffers(remove_duplicate=False)}
_stateless.functional_call(mod, params_and_buffers, args, kwargs)
```
is identical to calling the original module's forwards pass.

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse SciPioneer H-Huang

Pull Request resolved: https://github.com/pytorch/pytorch/pull/71542

Reviewed By: jbschlosser, albanD

Differential Revision: D33716716

Pulled By: Chillee

fbshipit-source-id: 8ae09a9fb8bf69d08d9a23cc3bc7e61c4ed826b3

albanD commented:
@Chillee what happened here?
